cartographer/common/lockless_queue.h

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2018 The Cartographer Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CARTOGRAPHER_COMMON_LOCKLESS_QUEUE_H_
+#define CARTOGRAPHER_COMMON_LOCKLESS_QUEUE_H_
+
+#include <atomic>
+#include <list>
+#include <memory>
+
+#include "cartographer/common/make_unique.h"
+#include "glog/logging.h"
+
+namespace cartographer {
+namespace common {
+
+// Lock-less queue which is thread safe for concurrent data producers and a
+// single data consumer.
+//
+// This lockless queue implementation is adapted from
+// https://github.com/resonance-audio/resonance-audio/blob/master/resonance_audio/utils/lockless_task_queue.h
+template <typename T>
+class LocklessQueue {
+ public:
+  LocklessQueue() {
+    free_list_head_ = nullptr;
+    incoming_data_list_head_ = nullptr;
+    data_list_head_ = nullptr;
+    data_list_tail_ = nullptr;
+  }
+
+  ~LocklessQueue() {
+    FreeNodes(free_list_head_.exchange(nullptr));
+    FreeNodes(incoming_data_list_head_.exchange(nullptr));
+    FreeNodes(data_list_head_);
+  }
+
+  // Pushes the data item into the queue.
+  void Push(std::unique_ptr<T> t) {
+    Node* const free_node = PopNodeFromFreeList();
+    CHECK(free_node);
+    free_node->data = std::move(t);
+    PushNodeToList(&incoming_data_list_head_, free_node);
+  }
+
+  // Pops the oldest data item from the queue. If the queue is empty returns a
+  // nullptr.
+  std::unique_ptr<T> Pop() {
+    PopAllDataNodes();
```

Comment:
Maybe rename this method to SwapLists, since this method actually doesn't execute pop operations.